### DISCUSSION TITLE
rrdtool TOTAL CF function

### DISCUSSION THREAD

**Discussion by axet** [oetiker/rrdtool-1.x#1232] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

rrdtool have 4 consolidation functions which are: AVERAGE, MIN, MAX, LAST. But when you need to create total sum of a values, you have to use average which is not accurate.

**Describe the solution you'd like**

New consolidation function TOTAL

**Describe alternatives you've considered**

Using AVERAGE + (MIN-MAX)

**Comment by oetiker** [comment]
Quoting from the rrdgraph_rpn man page

```
TOTAL

    Returns the rate from each defined time slot multiplied with the step size. This can, 
    for instance, return total bytes transferred when you have logged bytes per second. 
    The time component returns the number of seconds.

    Example: VDEF:total=mydata,TOTAL
```

**Reply by axet** [reply]
I know the doc, maybe I do not understand it. rrdtool writes data using consolidation function. What data source should I use for ds0? Since it can be only from 4 CF I have to use AVERAGE as ds0, right? I read the doc for  'rrdtool create' it seems CF function is mandatory. Regardless if this is mandatory or not, not using CF function can cause rrd file to grow up quickly. That is why I proposed CF_TOTAL

Source can be defined as:

```
DEF:ds0=/var/lib/munin/localdomain/localhost.localdomain-diskstats_throughput-mapper_luks_wrbytes-g.rrd:42:TOTAL
```

But here is no TOTAL as CF name

**Reply by oetiker** [reply, marked as answer]
no there is no TOTAL consolidation function ... as such a thing does not make any sense ... you would waste valuable space for useless data ... noone is interested in a random total ... people want TOTAL over a particular time periode ... to be able to calculate this you best store the average ... and then calculate the TOTAL over the timerange you are interested in.

the average rrdtool stores in a slot is TOTAL/duration

**Reply by axet** [reply]
I got it. This field AVERAGE is not a math definition of average / mean. It is rate composed of two values TOTAL and duration.

**Reply by oetiker** [reply]
It is a math definition "TOTAL divided by duration". The code takes great pains to properly calculate the AVERAGE ... it works!

**Reply by axet** [reply]
Then I do not understand it again. Is AVERAGE rrd using scalar (one number) or vector (two values) value?

**Reply by oetiker** [reply]
rrdtool integrates the space under the staircase curve built by the values you feed it. This number is then divided by 'size' of the slot you are storing the data in ... as mentioned the math is sound and there should not be any significant loss of precision.

**Reply by axet** [reply]
From scalar AVERAGE value you can not restore TOTAL value. Here: http://rrdtool.vandenbogaerdt.nl/min-avg-max.php or here http://rrdtool.vandenbogaerdt.nl/process.php it is saying "AVERAGE interval", which means AVERAGE is vector value. If so, then I understand how you recover TOTAL from AVERAGE. But hen AVERAGE is not linear algebra AVERAGE. That why I said above it is not a math term.

**Reply by oetiker** [reply]
We always store a 'rate'

Time 100: Counter = 40
Time 600: Counter = 90

Interval = 500
Counter difference = 50

Rate = 0.1

**Reply by oetiker** [reply]
the rrdtool code is written in a way that it can deal with counter values shipped at any point in time and still come up with the correct rate so that you can calculate the counter difference between any to points in time, assuming linear change of the counter between the readings
